include/Dialects/LinalgTransform/TransformOpInterface.h

```diff
@@ -0,0 +1,226 @@
+  /// yet.
+  constexpr const static unsigned kUnsetRangeStart = -1u;
+
+  /// Creates an instance of TransformResults that expects mappings for
+  /// `numSegments` values.
+  explicit TransformResults(unsigned numSegments);
+
+  /// Gets the list of operations associated with the result at the given
+  /// position.
+  ArrayRef<Operation *> get(unsigned position) const;
+
+  /// Storage for pointers to payload IR ops that are associated with results of
+  /// a transform IR op. `segments` contains as many entries as the transform IR
+  /// op has results. Each entry identifies a start and a length of a segment in
+  /// the `operations` list that contains the pointers to operations. This
+  /// allows for operations to be stored contiguously without nested vectors and
+  /// for different segments to be set in any order.
+  struct Range {
+    Range(unsigned start, unsigned length) : start(start), length(length) {}
+    unsigned start;
+    unsigned length;
+  };
+  SmallVector<Range, 2> segments;
+  SmallVector<Operation *> operations;
+};
+
+/// Applies a one-to-one transform to each of the given targets. Puts the
+/// results of transforms in `results` in the same order. Fails if any of the
+/// application fails. Individual transforms must be callable with signature
+///   FailureOr<convertible-to-Operation*>(OpTy)
+/// where OpTy is either
+///   - Operation *, in which case the transform is always applied;
+///   - a concrete Op class, in which case a check is performed whether
+///   `targets` contains operations of the same class and a failure if reported
+///   if it does not.
+template <typename FnTy>
+LogicalResult applyTransformToEach(ArrayRef<Operation *> targets,
+                                   SmallVectorImpl<Operation *> &results,
+                                   FnTy transform) {
+  using TransformOpType =
+      typename llvm::function_traits<FnTy>::template arg_t<0>;
+  for (Operation *target : targets) {
+    auto specificOp =
+        functional::detail::IsaOr<TransformOpType>::dyn_cast(target);
+    if (!specificOp)
+      return failure();
+
+    auto result = transform(specificOp);
+    if (failed(result))
+      return failure();
+
+    results.push_back(*result);
+  }
+  return success();
+}
+
+/// Trait implementing the TransformOpInterface for operations applying a
+/// transformation to a single operation handle and producing a single operation
+/// handle. The op must implement:
+///   - FailureOr<convertible-to-Operation*> applyToOne(OpTy) - a transformation
+///   that is applied in turn to all payload IR operations that correspond to
+///   the handle of the transform IR operation.
+/// In the functions above, OpTy is either Operation * or a concrete payload IR
+/// Op class that the transformation is applied to (NOT the class of the
+/// transform IR op).
+template <typename OpTy>
+class TargetableOpTrait : public OpTrait::TraitBase<OpTy, TargetableOpTrait> {
```

Comment:
nit: should `SingleOperandSingleResult` or something similar appear in the trait name ?